gmp-droid.cpp

```diff
@@ -0,0 +1,559 @@
+
+#include <iostream>
+#include <cstring>
+#include <map>
+#include <stdlib.h>
+#include <arpa/inet.h>
+
+#include "droidmediacodec.h"
+
+#include "gmp-platform.h"
+#include "gmp-video-host.h"
+
+#include "gmp-video-decode.h"
+#include "gmp-video-frame-i420.h"
+#include "gmp-video-frame-encoded.h"
+#include "gmp-droid-conv.h"
+#include "gmp-task-utils.h"
+
+#define CRITICAL 0
+#define ERROR 1
+#define INFO  2
+#define DEBUG 3
+
+const char *kLogStrings[] = {
+  "GMP-DROID Critical: ",
+  "GMP-DROID Error: ",
+  "GMP-DROID Info: ",
+  "GMP-DROID Debug: "
+};
+
+static int g_log_level = INFO;
+
+#define LOG(l, x) do { \
+        if (l >=0 && l <= g_log_level) { \
+            std::cerr << kLogStrings[l] << x << std::endl; \
+        } \
+    } while(0)
+
+static GMPPlatformAPI *g_platform_api = NULL;
+
+// Droidmedia callbacks
+static void DroidError (void *data, int err);   // pass error and abord
+static int SizeChanged (void *data, int32_t width, int32_t height);     // reconfigure
+static void DataAvailable (void *data, DroidMediaCodecData * decoded);
+static void SignalEOS (void *data);
+
+class DroidVideoDecoder : public GMPVideoDecoder
+{
+public:
+  explicit DroidVideoDecoder (GMPVideoHost * hostAPI):
+      m_host (hostAPI),
+      m_callback (NULL)
+  {
+    // Create the Mutex
+    GMPErr err = g_platform_api->createmutex (&m_codec_lock);
+    if (GMP_FAILED (err))
+        Error (err);
+    err = g_platform_api->createmutex (&m_stop_lock);
+    if (GMP_FAILED (err))
+        Error (err);
+  }
+
+  virtual ~DroidVideoDecoder ()
+  {
+    // Destroy the Mutex
+    m_codec_lock->Destroy ();
+    m_stop_lock->Destroy ();
+  }
+
+// GMPVideoDecoder methods
+  virtual void InitDecode (const GMPVideoCodec & codecSettings,
+      const uint8_t * aCodecSpecific,
+      uint32_t aCodecSpecificSize,
+      GMPVideoDecoderCallback * callback, int32_t coreCount)
+  {
+
+    m_callback = callback;
+
+    // Check if this device supports the codec we want
+    memset (&m_metadata, 0x0, sizeof (m_metadata));
+    m_metadata.parent.flags =
+        static_cast <DroidMediaCodecFlags> (DROID_MEDIA_CODEC_HW_ONLY | DROID_MEDIA_CODEC_NO_MEDIA_BUFFER);
+
+    switch (codecSettings.mCodecType) {
+      case kGMPVideoCodecVP8:
+        m_metadata.parent.type = "video/x-vnd.on2.vp8";
+        break;
+      case kGMPVideoCodecVP9:
+        m_metadata.parent.type = "video/x-vnd.on2.vp9";
+        break;
+      case kGMPVideoCodecH264:
+        m_metadata.parent.type = "video/avc";
+        break;
+      default:
+        LOG (ERROR, "Unknown GMP codec");
+        Error (GMPNotImplementedErr);
+        return;
+    }
+
+    // Check that the requested codec is actually available on this device
+    if (!droid_media_codec_is_supported (&m_metadata.parent, false)) {
+      LOG (ERROR, "Codec not supported");
+      Error (GMPNotImplementedErr);
+    }
+    // Set codec parameters
+    if (codecSettings.mWidth && codecSettings.mHeight) {
```

Comment:
What happens if either is zero?

Reply:
I'll remove that check.

Reply:
Done

Reply:
Is that wise?

Reply:
I checked the openh264 implementation and it doesn't check either. A video with 0 for the width or height is pretty much broken anyway. I guess it could be overridden in the frame data too.